### PR TITLE
Add support for BGP Asn Notation

### PR DESCRIPTION
**Pull request by eddyatdell**
Add support for BGP asn/as-dot notation (rfc-5395).

https://github.com/ansible-network/resource_module_models/pull/269

##### SUMMARY
- add "as_notation" attribute to BGP globals argspec to set the as-notation format for Bgp instance
- all Bgp AS numbers should be type str (from int) to support as-dot string format
- new classes to support asn compares between asdot+ (with ".") and asplain (no ".")
- expand regression and unit test cases to test the new functionality

##### GitHub Issues
| GitHub Issue # |
| -------------- |
|N/A |

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME

sonic_bgp
sonic_bgp_af
sonic_bgp_neighbors
sonic_bgp_neighbors_af
sonic_route_maps

##### OUTPUT

[bgp_asn_notation_regression_2024_08_02_22_40_54.zip](https://github.com/user-attachments/files/16476034/bgp_asn_notation_regression_2024_08_02_22_40_54.zip)

[bgp_asn_notation_regression.zip](https://github.com/user-attachments/files/16254818/bgp_asn_notation_regression.zip)

##### Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [x] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have maintained backward compatibility or have provided any relevant "breaking_changes" descriptions in a "fragment" file in the "changelogs/fragments" directory of this repository.
- [x] I have provided a summary for this PR in valid "fragment" file format in the "changelogs/fragments" directory of this repository branch. Reference : [Ansible Change Log Document](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to)

##### How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- verified UT and sanity test results
- verified regression tests (attached)

### CODE REVIEW

File: plugins/module_utils/network/sonic/utils/bgp_utils.py [hunk 1]
```diff
@@ -28,11 +28,195 @@
...
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def to_request(self):
+        """Return asn according to openconfig model (original input: asdot(+) as string, asplain as integer)"""
+        return self.intval if self.__str__().find('.') < 0 else self.__str__()
+
+
+def get_bgp_asn(cfglist):
```
Comment (mingjunzhang2019) [inline plugins/module_utils/network/sonic/utils/bgp_utils.py:89]: 1, This function does not return value. So suggest to change the name to like transform_*** or convert_*** or something.
2, It handles both bgp and bgp_neighbors, so may I suggest to move it to individual config file to do the function specifically.

**Comment by mingjunzhang2019** [comment]
Now bgp AS changed from int to str. Should this case be handled or is  it handled: if target SONIC system has have bgp configuration with bgp AS as int and new playbook has bgp AS as str? I see class BgpAsn handles the case AS is int.

**Comment by eddyatdell** [comment]
> Now bgp AS changed from int to str. Should this case be handled or is it handled in the code: if target SONIC system has have bgp configuration with bgp AS as int and new playbook has bgp AS as str? I see class BgpAsn handles the case AS is int.

The BGP Asn Notation feature is a enterprise-sonic release 4.4 feature. The as_notation attribute and asdot+ strings won't work on older releases.

File: plugins/module_utils/network/sonic/utils/bgp_utils.py (hunk 1 above)
Comment (eddyatdell) [reply plugins/module_utils/network/sonic/utils/bgp_utils.py]: 1. Rename conversion functions to convert_xxx().
2. Other than the global sonic_bgp, the bgp_af, bgp_neighbors and bgp_neighbors_af have both global and neighbor configs that need to be converted.

Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/utils/bgp_utils.py]: Thanks for comments.

**Comment by mingjunzhang2019** [comment]
> > Now bgp AS changed from int to str. Should this case be handled or is it handled in the code: if target SONIC system has have bgp configuration with bgp AS as int and new playbook has bgp AS as str? I see class BgpAsn handles the case AS is int.
> 
> The BGP Asn Notation feature is a enterprise-sonic release 4.4 feature. The as_notation attribute and asdot+ strings won't work on older releases.

Thanks for the comment.

**Review by mingjunzhang2019 — APPROVED** [review]

**Review by kerry-meyer — CHANGES_REQUESTED** [review]
The proposed code changes and corresponding test results posted for this PR look good.

While the code logic and structure look fine, I am requesting a change in the name of the 'to_request' class methods proposed here to avoid confusion for code maintainers when browsing or searching the code. (See details in my comment for the 'bgp_utils.py' file.)

**Comment by kerry-meyer** [comment]
I have an additional request: Please resolve the conflicts for this PR to prepare the PR for merging.

File: plugins/modules/sonic_bgp.py [hunk 2]
```diff
@@ -65,6 +65,11 @@
         description:
           - Enables/disables logging neighbor up/down and reset reason.
         type: bool
+      as_notation:
+        description:
```
Comment (kerry-meyer) [inline plugins/modules/sonic_bgp.py:69]: Please add a line to this description stating that this option is only supported for SONiC releases at or after 4.4.0.

File: plugins/module_utils/network/sonic/utils/bgp_utils.py
[the remark this answers is not in the document]
Comment (eddyatdell) [reply plugins/module_utils/network/sonic/utils/bgp_utils.py:1]: Done. Btw, these methods are also doing conversions for out-going REST-API.

File: plugins/modules/sonic_bgp.py (hunk 2 above)
Comment (eddyatdell) [reply plugins/modules/sonic_bgp.py:69]: Done.  New regression results in description: bgp_asn_notation_regression_2024_08_02_22_40_54.zip

File: plugins/module_utils/network/sonic/utils/bgp_utils.py
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/utils/bgp_utils.py:1]: Thank you for making these changes.

(These methods are doing a conversion for a single attribute within a request as opposed to doing the conversion for the entire request. That was the reason for requesting this name change.)

Resolved.

File: plugins/modules/sonic_bgp.py (hunk 2 above)
Comment (kerry-meyer) [reply plugins/modules/sonic_bgp.py:69]: Thank you for adding this explanation.

Resolved.

**Review by kerry-meyer — APPROVED** [review]
The current set of proposed code changes and corresponding test results look good.

Thank you for providing this enhancement to the capabilities of the enterprise_sonic Ansible resource module collection.

Approved.

**Pull request merged**